product_packaging_level/models/product_packaging.py

```diff
@@ -0,0 +1,155 @@
+    packaging_level_id = fields.Many2one(
+        "product.packaging.level",
+        required=True,
+        ondelete="restrict",
+        default=lambda p: p.default_packaging_level_id(),
+    )
+    barcode_required_for_gtin = fields.Boolean(
+        readonly=True, compute="_compute_barcode_required_for_gtin"
+    )
+    qty_per_level = fields.Char(
+        compute="_compute_qty_per_level", string="Qty per package level"
+    )
+    name_policy = fields.Selection(related="packaging_level_id.name_policy")
+    sequence = fields.Integer(compute="_compute_sequence", store=True, readonly=False)
+
+    def default_packaging_level_id(self):
+        return self.env["product.packaging.level"].search(
+            [("is_default", "=", True)], limit=1
+        )
+
+    @api.constrains("packaging_level_id", "product_id")
+    def _check_one_packaging_level_per_product(self):
+        packages_name_by_level = self.filtered(
+            lambda p: p.name_policy == "by_package_level"
+        )
+        for packaging in packages_name_by_level:
+            product = packaging.product_id
+            # do not use a mapped/filtered because it would union the duplicates
+            packaging_level_ids = [
+                packaging.packaging_level_id.id
+                for packaging in product.packaging_ids
+                # We have to allow several packaging using the default type,
+                # because when we install the module on an existing database,
+                # the default value will be set to default and we'll have
+                # duplicates. Anyway "default" is not meant to be used as a
+                # real type.
+                if not packaging.packaging_level_id.is_default
+            ]
+            if len(set(packaging_level_ids)) != len(packaging_level_ids):
+                raise ValidationError(
+                    _(
+                        "It is forbidden to have different packagings "
+                        "with the same level for a given product ({})."
+                    ).format(product.display_name)
+                )
+
+    @api.depends("packaging_level_id")
+    def _compute_sequence(self):
+        for packaging in self:
+            packaging.sequence = packaging.packaging_level_id.sequence
+
+    @api.depends("packaging_level_id", "packaging_level_id.has_gtin", "qty")
+    def _compute_barcode_required_for_gtin(self):
+        for packaging in self:
+            packaging.barcode_required_for_gtin = packaging.packaging_level_id.has_gtin
+            if not packaging.qty:
+                packaging.barcode_required_for_gtin = False
+
+    @api.depends(
+        "product_id",
+        "product_id.packaging_ids",
+        "packaging_level_id",
+        "packaging_level_id.code",
+    )
+    def _compute_qty_per_level(self):
+        for packaging in self:
+            if not packaging.product_id:
+                packaging.qty_per_level = ""
+                continue
```

Comment:
As `product_id` is a required field starting from v17, these lines could be removed.